### PR TITLE
[Improve](InPredict) enhance in predict with struct type

### PR DESCRIPTION
**Pull request by amorynan**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/functions/function_struct_in.h [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        return column->compare_at(row_idx, other.row_idx, *column, 0) == 0;
+    }
+    // compare
+    bool operator<(const ColumnRowRef& other) const {
+        return column->compare_at(row_idx, other.row_idx, *column, 0) < 0;
+    }
+
+    // when call set find, will use hash to find
+    size_t operator()(const ColumnRowRef& a) const {
+        std::vector<uint32_t> hash_vec(1);
```
Comment (HappenLee) [inline be/src/vec/functions/function_struct_in.h:59]: why use vector，not `uint32_t`

File: be/src/vec/functions/function_struct_in.h [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            if (const_column_ptr == nullptr) {
+                break;
+            }
+            const auto& [col, _] = unpack_if_const(const_column_ptr->column_ptr);
+            if (col->is_nullable()) {
+                auto* null_col = vectorized::check_and_get_column<vectorized::ColumnNullable>(col);
+                if (!null_in_set && null_col->has_null()) {
+                    null_in_set = true;
+                }
+                column_struct_ptr_args->insert_from(null_col->get_nested_column(), 0);
```
Comment (HappenLee) [inline be/src/vec/functions/function_struct_in.h:109]: if the value is NULL, no need call `insert_from`

File: be/src/vec/functions/in.h [hunk 3]
```diff
@@ -69,6 +70,7 @@ template <bool negative>
 class FunctionIn : public IFunction {
 public:
     static constexpr auto name = negative ? "not_in" : "in";
+    std::unique_ptr<vectorized::Arena> _predicate_arena = std::make_unique<vectorized::Arena>();
```
Comment (HappenLee) [inline be/src/vec/functions/in.h:73]: remove the arena logic

File: be/src/vec/functions/function_struct_in.h (hunk 2 above)
Comment (amorynan) [reply be/src/vec/functions/function_struct_in.h:109]: make sense! here need else

File: be/src/vec/functions/function_struct_in.h (hunk 1 above)
Comment (amorynan) [reply be/src/vec/functions/function_struct_in.h:59]: I tried use uint32_t , but I make some tests , but got different val in hash_val

**Comment by amorynan** [comment]
run buildall

**Review by HappenLee — DISMISSED** [review]
LGTM

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/TypeCoercionUtils.java [hunk 4]
```diff
@@ -962,7 +963,13 @@ public static Expression processInPredicate(InPredicate inPredicate) {
                         .stream()
                         .map(Expression::getDataType).collect(Collectors.toList()),
                 true);
-        if (optionalCommonType.isPresent() && !supportCompare(optionalCommonType.get())) {
+        if (inPredicate.getCompareExpr().getDataType().isStructType() && optionalCommonType.isPresent()
+                && !optionalCommonType.get().isStructType()) {
+            throw new AnalysisException("data type " + optionalCommonType.get()
+                    + " could not used in InPredicate " + inPredicate.toSql());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/util/TypeCoercionUtils.java:969]: this exception message is inappropriate

**Comment by amorynan** [comment]
run buildall

**Comment by amorynan** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**